// appium-xcuitest-driver/build/lib/wda/iproxy.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.iProxy = void 0;

require("source-map-support/register");

var _net = _interopRequireDefault(require("net"));

var _appiumSupport = require("appium-support");

var _appiumIosDevice = require("appium-ios-device");

const log = _appiumSupport.logger.getLogger('iProxy');

class iProxy {
  constructor(udid, localport, deviceport) {
    this.localport = parseInt(localport, 10);
    this.deviceport = parseInt(deviceport, 10);
    this.udid = udid;
    this.serverSocket = null;
  }

  start() {
    if (this.serverSocket) {
      return;
    }

    this.serverSocket = _net.default.createServer(async connection => {
      try {
        const socket = await _appiumIosDevice.utilities.connectPort(this.udid, this.deviceport);
        socket.on('close', connection.destroy);
        socket.on('error', log.error);
        connection.on('close', socket.destroy);
        connection.on('error', log.error);
        connection.pipe(socket);
        socket.pipe(connection);
      } catch (e) {
        log.warn(e.message);
        connection.destroy();
      }
    });
    this.serverSocket.listen(this.localport);
  }

  quit() {
    if (!this.serverSocket) {
      return;
    }

    this.serverSocket.close();
    this.serverSocket = null;
  }

}

exports.iProxy = iProxy;
var _default = iProxy;
exports.default = _default;require('source-map-support').install();
