// appium-xcuitest-driver/build/lib/device-log/ios-device-log.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.IOSDeviceLog = void 0;

require("source-map-support/register");

var _appiumSupport = require("appium-support");

var _iosLog = require("./ios-log");

var _appiumIosDevice = require("appium-ios-device");

const log = _appiumSupport.logger.getLogger('IOSDeviceLog');

class IOSDeviceLog extends _iosLog.IOSLog {
  constructor(opts) {
    super();
    this.udid = opts.udid;
    this.showLogs = !!opts.showLogs;
    this.service = null;
  }

  async startCapture() {
    if (this.service) {
      return;
    }

    this.service = await _appiumIosDevice.services.startSyslogService(this.udid);
    this.service.start(this.onLog.bind(this));
  }

  onLog(logLine) {
    this.broadcast(logLine);

    if (this.showLogs) {
      log.info(logLine);
    }
  }

  get isCapturing() {
    return !!this.service;
  }

  stopCapture() {
    if (!this.service) {
      return;
    }

    this.service.close();
    this.service = null;
  }

}

exports.IOSDeviceLog = IOSDeviceLog;
var _default = IOSDeviceLog;
exports.default = _default;require('source-map-support').install();
